### PR TITLE
messenger: use async, allow child to return before backoff completes

### PR DESCRIPTION
**Pull request by cole-h**
##### Description

Prior to this change, if the vault agent failed before backoff
completed, we would wait until the backoff completed and _then_ fail.
Since the backoff maxes out at 15 minutes, it's better to fail the unit
earlier.

Rather than dealing with threads (manually, at least), we use async-std
in order to start running the backoff in the background, allowing us to
wait on the vault agent in the foreground and retrieve its status code
if it exists with a failure.

##### Checklist

- [x] Formatted with `nixpkgs-fmt`
- [x] Ran tests with:
  * `nix-instantiate --strict --eval --json ./default.nix -A checks.definition`
  * `nix-instantiate --strict --eval --json ./default.nix -A checks.helpers`
  * `nix-build ./default.nix -A checks.implementation`
  * `cargo test --manifest-path ./messenger/Cargo.toml`
- [x] Added or updated relevant tests (leave unchecked if not applicable)
- [ ] Added or updated relevant documentation (leave unchecked if not applicable)

### CODE REVIEW

**Review by grahamc — COMMENTED** [review]
lgtm but I'd love @firestack to review.

File: messenger/src/main.rs [hunk 1]
```diff
@@ -50,35 +54,64 @@ fn main() -> Result<()> {
...
+                    },
+                    complete => break,
+                };
             }
         }
         Err(err) => {
             error!("failed to spawn vault agent with args: {:?}", command);
-            error!("{:?}", err);
-            sd_notify::notify(false, &[NotifyState::Errno(1)])?;
+            error!("{}", err);
```
Comment (firestack) [inline messenger/src/main.rs:113]:
```suggestion
            error!(%err);
```

File: messenger/src/main.rs [hunk 2]
```diff
@@ -50,35 +54,64 @@ fn main() -> Result<()> {
...
+                            errno
+                        } else if let Some(signal) = status.signal() {
+                            status_msg.push_str(&format!(" with signal {}", signal));
+                            signal
+                        } else {
+                            status_msg.push_str(" with unknown cause (not exit code or signal)");
+                            1
+                        };
+
+                        error!("{}", status_msg);
```
Comment (firestack) [inline messenger/src/main.rs:103]:
```suggestion
                        error!(%status_msg);
```

File: messenger/src/main.rs [hunk 3]
```diff
@@ -50,35 +54,64 @@ fn main() -> Result<()> {
...
+        .with(env_filter)
+        .with(crate_filter)
         .try_init()?;
 
     let mut command = Command::new(cli.vault_binary);
     command.arg("agent");
     command.arg("-config");
     command.arg(cli.agent_config);
 
+    trace!("spawning vault agent with args: {:?}", command);
```
Comment (firestack) [inline messenger/src/main.rs:66]:
```suggestion
    trace!(?command, "spawning vault agent", );
```

File: messenger/src/main.rs [hunk 4]
```diff
@@ -50,35 +54,64 @@ fn main() -> Result<()> {
...
+            let backoff = self::backoff_until_files_exist(files).fuse();
+            let status = child.status().fuse();
+
+            futures::pin_mut!(backoff, status);
+
+            loop {
+                futures::select! {
+                    backoff = backoff => {
+                        if let Err(err) = backoff {
+                            error!("backoff failed: {}", err);
```
Comment (firestack) [inline messenger/src/main.rs:83]:
```suggestion
                            error!(%err, "backoff failed");
```

**Review by firestack — APPROVED** [review]
LGTM!

I went and made a few suggestions from what I know about tracing and structured logging in tracing, feel free to ignore any and all of them.

**Comment by cole-h** [comment]
Applied your suggestions re: tracing. Thanks!

**Pull request merged**